app/Http/Controllers/Webapp/PostController.php

```diff
@@ -0,0 +1,1890 @@
+
+        $mess_error = Post::checkPostDataRequireField($data, true);
+
+        if($mess_error != ''){
+            $this->_flashMessage($request, $mess_error, 'danger');
+            $this->activityLog->setInfo('Missing required field');
+            return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+        }
+
+        if($post->site_id == 0) {
+            $this->_flashMessage($request, 'ターゲットサイトを選んでください', 'warning');
+            $this->activityLog->setInfo('Missing target site');
+            if(!$request->get('selected_ids'))
+                return redirect()->to(action('Webapp\\PostController@getEdit', $postId))->send();
+            else
+                return;
+        }
+
+        if (!Post::isUniquePostByTitle($post)) {
+            $this->_flashMessage($request, 'このタイトルは存在しています。公開できないのでタイトルを変更してください。', 'danger');
+            $this->activityLog->addMeta('Error', 'Title duplicated: '.$post->title);
+            if(!$request->get('selected_ids'))
+                return redirect()->to(action('Webapp\\PostController@getIndex'))->send();
+            else
+                return;
+        }
+
+        try {
+            $apiUrl = Site::find($post->site_id)->api_url;
+            $action = ($post->target_post_id == 0) ? 'publish' : 'edit';
+            $returnData = (new CdsApiDriver($apiUrl, [
+                'content' => $post->content,
+                'title' => $post->title,
+                'categories' => [$post->category],
+                'post_id' => $post->target_post_id,
+                'post_excerpt' => $post->description,
+                'post_status' => 'publish',
+                'feature_img' => $post->feature_img,
+                'author_name' => $post->author_name,
+                'cback_url' => action('Webapp\\PostController@getEdit', $post->id),
+                'priority' => $post->priority,
+            ]))->send($action)->getResult();
+            $post->content = Post::replaceIDButtonLink($returnData->data->post_id, $post->content, $post->target_post_id);
+            $post->target_post_id = $returnData->data->post_id;
+            $post->published_status = 1;
+            $this->_flashMessage($request, 'OK', 'success');
+
+            $this->activityLog->addMeta('TargetPostId', $post->target_post_id);
+        } catch(\Exception $e) {
+            $post->content = Post::replaceIDButtonLink('__POSTID__', $post->content, $post->target_post_id);
+            $this->activityLog->setInfo('Error when reaching to target site');
+            $this->activityLog->addMeta('FailedPostToTargetSite', 'error when reaching to target site');
+            $this->_flashMessage($request, '更新に失敗しました。もう一度ご確認ください。', 'warning');
+        }
+
+        $post->save();
+
+        $this->activityLog->addMeta('NewPublishedStatus', $post->published_status);
+
+        $next = action('Webapp\\PostController@getIndex');
+
+        if(($from = $request->get('from', null) != null) && $from == 'preview') {
```

Comment:
![CRITICAL](https://sonarsource.github.io/sonar-github/severity-critical.png 'Severity: CRITICAL') Define a constant instead of duplicating this literal "preview" 3 times. [![rule](https://sonarsource.github.io/sonar-github/rule.png)](http://sonar.relipa.blog/coding_rules#rule_key=php%3AS1192)